src/FastMathFunctions.cpp

```diff
@@ -0,0 +1,872 @@
+                prec.constraint_max_absolute_error = iida.default_mae;
+            }
+        }
+    }
+    return ii;
+}
+
+bool intrinsic_satisfies_precision(const IntrinsicsInfo &ii, const ApproximationPrecision &prec) {
+    if (!ii.intrinsic.defined()) {
+        return false;
+    }
+    if (prec.force_halide_polynomial) {
+        return false;  // Don't use intrinsics if the user really wants a polynomial.
+    }
+    if (prec.optimized_for != ii.intrinsic.behavior) {
+        return false;
+    }
+    if (prec.constraint_max_ulp_error != 0) {
+        if (ii.intrinsic.max_ulp_error != 0) {
+            if (ii.intrinsic.max_ulp_error > prec.constraint_max_ulp_error) {
+                return false;
+            }
+        } else {
+            // We don't know?
+        }
+    }
+    if (prec.constraint_max_absolute_error != 0) {
+        if (ii.intrinsic.max_abs_error != 0) {
+            if (ii.intrinsic.max_abs_error > prec.constraint_max_absolute_error) {
+                return false;
+            }
+        } else {
+            // We don't know?
+        }
+    }
+    return true;
+}
+
+bool native_func_satisfies_precision(const IntrinsicsInfo &ii, const ApproximationPrecision &prec) {
+    if (prec.force_halide_polynomial) {
+        return false;  // Don't use native functions if the user really wants a polynomial.
+    }
+    if (!ii.native_func.defined()) {
+        return true;  // Unspecified means it's exact.
+    }
+    if (prec.optimized_for != ii.native_func.behavior) {
+        return false;
+    }
+    if (prec.constraint_max_ulp_error != 0) {
+        if (ii.native_func.max_ulp_error != 0) {
+            if (ii.native_func.max_ulp_error > prec.constraint_max_ulp_error) {
+                return false;
+            }
+        } else {
+            // We don't know?
+        }
+    }
+    if (prec.constraint_max_absolute_error != 0) {
+        if (ii.native_func.max_abs_error != 0) {
+            if (ii.native_func.max_abs_error > prec.constraint_max_absolute_error) {
+                return false;
+            }
+        } else {
+            // We don't know?
+        }
+    }
+    return true;
+}
+
+class LowerFastMathFunctions : public IRMutator {
+    using IRMutator::visit;
+
+    const Target &target;
+    DeviceAPI for_device_api = DeviceAPI::None;
+
+    bool is_cuda_cc20() {
+        return for_device_api == DeviceAPI::CUDA && target.get_cuda_capability_lower_bound() >= 20;
+    }
+    bool is_cuda_cc75() {
+        return for_device_api == DeviceAPI::CUDA && target.get_cuda_capability_lower_bound() >= 75;
+    }
+
+    void adjust_precision_for_target(ApproximationPrecision &prec) {
+        if (for_device_api == DeviceAPI::None) {
+            if (target.arch == Target::Arch::X86) {
+                // If we do not have fused-multiply-add, we lose some precision.
+                if (target.bits == 32 || !target.has_feature(Target::Feature::FMA)) {
```

Comment:
The risk here is that the target may not have been completed. e.g. avx2 might be set, but not fma, even though avx2 implies fma. We could expose complete_x86_target from CodeGen_X86.cpp. Really that's a whole separate PR though. Maybe just change this to `target.features_any_of({Target::Feature::FMA, Target::Feature::AVX2, Target::Feature::AVX512, Target::Feature::AVX10_1})` and add a TODO to revisit this when #7839 is addressed.

Reply:
Deleted that `adjust_precision_for_target` helper, as now all advertised precision is measured exhaustively without FMA.